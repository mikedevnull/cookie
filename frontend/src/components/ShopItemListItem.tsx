import { Item } from "@/db/types";
import {
  ListItem,
  ListItemButton,
  ListItemIcon,
  Checkbox,
  ListItemText,
} from "@mui/material";
import { useEffect, useId, useState } from "react";

type Props = {
  name: string;
  active: boolean;
  onToggle?: () => void;
};

const TOGGLE_DELAY_IN_MS = 1000 as const;

type ItemState = "active" | "transitioning" | "inactive";

export default function ShopItemListItem({ name, active, onToggle }: Props) {
  const [itemState, setItemState] = useState<ItemState>(
    active ? "active" : "inactive"
  );
  useEffect(() => {
    if (itemState === "transitioning") {
      const timer = setTimeout(() => {
        if (onToggle) {
          onToggle();
        }
      }, TOGGLE_DELAY_IN_MS);
      return () => {
        clearTimeout(timer);
      };
    }
  }, [itemState, { name, active }, onToggle]);

  const onClick = () => {
    if (itemState === "active") {
      setItemState("transitioning");
    } else {
      if (!active && onToggle) {
        onToggle();
      }
      setItemState("active");
    }
  };

  const labelId = useId();
  return (
    <ListItem disablePadding>
      <ListItemButton onClick={onClick}>
        <ListItemIcon>
          <Checkbox
            edge="start"
            checked={itemState !== "active"}
            tabIndex={-1}
            disableRipple
            inputProps={{ "aria-labelledby": labelId }}
          />
        </ListItemIcon>
        <ListItemText
          id={labelId}
          primary={name}
          primaryTypographyProps={
            itemState !== "active"
              ? {
                  style: { textDecoration: "line-through" },
                }
              : {}
          }
        />
      </ListItemButton>
    </ListItem>
  );
}
