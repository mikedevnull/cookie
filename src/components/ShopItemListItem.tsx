import { Paper } from "@mantine/core";
import classes from "./ShopItemListItem.module.css";

export type ShopItem = {
  name: string;
};

export default function ShopItemListItem({ item }: { item: ShopItem }) {
  let usedClasses = `${classes.item}`;
  const selected = true;
  if (selected) {
    usedClasses += ` ${classes.selected}`;
  }
  return (
    <Paper className={usedClasses} shadow="xs" p="xl">
      {item.name}
    </Paper>
  );
}
