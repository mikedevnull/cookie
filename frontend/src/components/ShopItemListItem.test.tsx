import { act, render, screen } from "@testing-library/react";
import ShopItemListItem from "./ShopItemListItem";
import userEvent, { UserEvent } from "@testing-library/user-event";

describe("ShopItemListItem", () => {
  let user: UserEvent;
  beforeEach(() => {
    jest.useFakeTimers();
    user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
  });
  afterEach(() => {
    act(() => jest.runOnlyPendingTimers());
    jest.useRealTimers();
  });

  it("Should render active items correctly", () => {
    const renderedItem = render(
      <ShopItemListItem item={{ name: "Foo", active: true }} />
    );
    expect(renderedItem).toMatchSnapshot();
  });

  it("Should render inactive items correctly", () => {
    const renderedItem = render(
      <ShopItemListItem item={{ name: "Foo", active: false }} />
    );
    expect(renderedItem).toMatchSnapshot();
  });

  it("Clicking an active item should change appearence to inactive immediately", async () => {
    render(<ShopItemListItem item={{ name: "Foo", active: true }} />);
    const listItem = await screen.findByText("Foo");
    await user.click(listItem);
    const checkBox = await screen.findByLabelText<HTMLInputElement>("Foo");
    expect(checkBox.checked).toBeTruthy();
  });

  it("Clicking an inactive item should change appearence to active immediately", async () => {
    render(<ShopItemListItem item={{ name: "Foo", active: false }} />);
    const listItem = await screen.findByText("Foo");
    await user.click(listItem);
    const checkBox = await screen.findByLabelText<HTMLInputElement>("Foo");
    expect(checkBox.checked).toBeFalsy();
  });

  it("Clicking an item should invoke toggle callback after delay", async () => {
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
    jest.useFakeTimers();
    const cb = jest.fn();
    render(
      <ShopItemListItem item={{ name: "Foo", active: true }} onToggle={cb} />
    );

    const listItem = await screen.findByText("Foo");
    await user.click(listItem);
    expect(cb).not.toHaveBeenCalled();

    act(() => jest.runOnlyPendingTimers());
    expect(cb).toHaveBeenCalled();
  });

  it("Clicking an item twice withing a short timefreame should invoke not invoke callback", async () => {
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
    jest.useFakeTimers();
    const cb = jest.fn();
    render(
      <ShopItemListItem item={{ name: "Foo", active: true }} onToggle={cb} />
    );

    const listItem = await screen.findByText("Foo");
    await user.click(listItem);
    act(() => jest.advanceTimersByTime(100));

    expect(cb).not.toHaveBeenCalled();

    await user.click(listItem);

    act(() => jest.runOnlyPendingTimers());
    expect(cb).not.toHaveBeenCalled();
  });
});
