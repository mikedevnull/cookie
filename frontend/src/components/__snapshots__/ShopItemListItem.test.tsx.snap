// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopItemListItem Should render active items correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-edgeStart MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-zkqepn-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                aria-labelledby=":r0:"
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                tabindex="-1"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            id=":r0:"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Foo
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-edgeStart MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-zkqepn-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              aria-labelledby=":r0:"
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          id=":r0:"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Foo
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ShopItemListItem Should render inactive items correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-checked PrivateSwitchBase-edgeStart MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-zkqepn-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                aria-labelledby=":r1:"
                checked=""
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                tabindex="-1"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </span>
          </div>
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            id=":r1:"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              style="text-decoration: line-through;"
            >
              Foo
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
    >
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-checked PrivateSwitchBase-edgeStart MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-zkqepn-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              aria-labelledby=":r1:"
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          id=":r1:"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            style="text-decoration: line-through;"
          >
            Foo
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
